Properly bootstrap Drupal when using an existing database

ChexWarrior opened the pull request:
<!--
 Before you open the request please review the following guidelines and tips to help it be more easily integrated:

 - Describe the scope of your change - i.e. what the change does.
 - Describe any known limitations with your change.
 - Please run any tests or examples that can exercise your modified code.

 Thank you for contributing! We will try to test and integrate the change as soon as we can, but be aware we have many GitHub repositories to manage and can't immediately respond to every request. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).

 Also don't be worried if the request is closed or not integrated sometimes the priorities of Bitnami might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
 -->

**Description of the change**

This PR is meant to fix the bug described in this [ticket](https://github.com/bitnami/charts/issues/5434) within the bitnami/charts repository.

Currently this PR is pretty rough/hacky, but I'm creating it now so I can start a discussion with the Bitnami team about the best way to fix this bug.  The original issue (linked above) is that when skipping the drupal bootstrap step the generated settings file doesn't have any of the information it needs to connect to the existing database and is missing some other important config information.  

**Current Approach**
As I worked on this I realized there were three items missing from the settings.php file when skipping the bootstrap step:
1. DB Connection Info (the $databases array)
2. The config sync directory (the $settings['config_sync_directory'] value)
3. The hash salt (the $settings['hash_salt'] value)

To add the DB connection info I just copied the drupal_set_database_ssl_settings (in libdrupal.sh) functions into a new function called drupal_set_database_settings and removed the SSL info for the pdo connection.  For the config_sync directory I just used the drupal_conf_set function to add the correct setting and I also had to create that directory ahead of time.  Right now its hardcoded, but it should be controllable by the user.  Finally, I used the drupal_conf_set function to create the hash salt, and again for now its hardcoded but it can be changed to something user controlled.  I'll keep working on this, but I'd like to hear if the Bitnami team thinks this is the proper direction to fix the bug.

**Applicable issues**

https://github.com/bitnami/charts/issues/5434

Thank you

alemorcuq commented:
Hi @ChexWarrior! I think this is the way to go for fixing this issue, so thank you very much for creating a PR!

Let me ask one question regarding your message: why remove the PDO settings for the database?

ChexWarrior commented:
@alemorcuq thank you for weighing in!  I removed the PDO settings mostly because I was rushing to get this to work, my understanding was that the method I copied, **drupal_set_database_ssl_settings**, added some specific PDO settings dealing with SSL as well as the rest of the db settings.  I wanted to create a method that would just add the base database settings, **drupal_set_database_settings**, so I removed the pdo key.

alemorcuq commented on 8/debian-10/rootfs/opt/bitnami/scripts/drupal-env.sh:
```diff
@@ -77,6 +77,8 @@ export DRUPAL_PROFILE="${DRUPAL_PROFILE:-standard}" # only used during the first
 export DRUPAL_SITE_NAME="${DRUPAL_SITE_NAME:-My blog}" # only used during the first initialization
 export DRUPAL_SKIP_BOOTSTRAP="${DRUPAL_SKIP_BOOTSTRAP:-}" # only used during the first initialization
 export DRUPAL_ENABLE_MODULES="${DRUPAL_ENABLE_MODULES:-}" # only used during the first initialization
+export DRUPAL_CONFIG_DIR="${DRUPAL_CONFIG_DIR:-sites/default/files/config}"
```
I think it's better to call this `DRUPAL_CONFIG_SYNC_DIR`. Why that default value, though? I can see that, on a new installation:

```
$settings['config_sync_directory'] = 'sites/default/files/config_Jwl9OpCE406LQ0XFpm5K_A_sh1kONuJ3hJpLA1N1BFRFFVXdEgeTgqiO0i_rt5cHcex5dQErVw/sync';
```

Being it a different path anytime you deploy a new installation.

ChexWarrior replied:
@alemorcuq I've renamed the var as you specified and updated the logic to create a "random" config directory if not value is specified by the user.  It follows the same logic as Drupal core using the Drupal\Component\Utility\Crypt::randomBytesBase64 method.

alemorcuq commented on 8/debian-10/rootfs/opt/bitnami/scripts/libdrupal.sh:
```diff
@@ -442,6 +458,21 @@ drupal_set_database_ssl_settings() {
 EOF
 }
 
+drupal_set_database_settings() {
+    cat >>"$DRUPAL_CONF_FILE" <<EOF
+\$databases['default']['default'] = array ( // Database block with SSL support
+  'database' => '${DRUPAL_DATABASE_NAME}',
+  'username' => '${DRUPAL_DATABASE_USER}',
+  'password' => '${DRUPAL_DATABASE_PASSWORD}',
+  'prefix' => '',
+  'host' => '${DRUPAL_DATABASE_HOST}',
+  'port' => '${DRUPAL_DATABASE_PORT_NUMBER}',
+  'namespace' => 'Drupal\\Core\\Database\\Driver\\mysql',
+  'driver' => 'mysql',
+);
+EOF
+}
```
I think we should also configure the PDO for SSL is the user is using that environment variable.

ChexWarrior replied:
I've updated the code so that if the $DRUPAL_DATABASE_TLS_CA_FILE var is specified it adds the SSL settings with the rest of the DB settings, otherwise it just adds the usual DB settings.

alemorcuq reviewed (commented):
Thanks. I'll try to test your changes as soon as I can.

ChexWarrior commented:
@alemorcuq Thank you for the feedback, I'll try and address your feedback by the end of the week.

alemorcuq commented:
Hi, @ChexWarrior. Sorry for the late response, I missed your request for review.

Your changes look good, but I have been testing them and it doesn't seem to work. I'm able to launch Drupal using an already existing and populated database with the `DRUPAL_SKIP_BOOTSTRAP` variable set to `yes`. The site loads, but I get a 404 error on resources such as CSS files. I imagine this is something that comes from the existing database and that has to be updated. Did you find the same issue when testing on your side?

ChexWarrior commented:
@alemorcuq Can you give anymore details on the error?  Is your site using a custom theme?

Here is the procedure I'm using to test the changes:
1. Spin up a mariadb container
2. Spin up a drupal container with `DRUPAL_SKIP_BOOTSTRAP` set to no
3. Ensure the site is working
4. Export a dump of the sites db
5. Destroy both the drupal and mariadb containers
6. Spin up mariadb with the previous drupal sites db in the `/docker-entrypoint-initdb.d` directory
7. Spin up drupal with `DRUPAL_SKIP_BOOTSTRAP` set to yes
8. Ensure the site is working

I'll try testing a site with a custom theme and resources to see if that is working as expected.

ChexWarrior commented:
I've tested with a simple custom theme and haven't run into any issues. Here were the steps I took:

1. Created a custom image (`nycedcmis/bitnami-drupal-custom-theme`) using  the bitnami-drupal image built from my fork as a base and then added in my simple custom  theme (called custom_theme) by copying it to `/opt/bitnami/drupal/themes/custom`.
2. Ran a mariadb container:
```
docker run --rm -d --name mariadb \
	--env ALLOW_EMPTY_PASSWORD=yes \
	--env MARIADB_USER=bn_drupal \
	--env MARIADB_PASSWORD=bitnami \
	--env MARIADB_DATABASE=bitnami_drupal \
	--network drupal-network \
	bitnami/mariadb:latest
```
3. Ran a drupal container using the aforementioned image, **w/o skipping the bootstrap**:
```
docker run --rm --name drupal \
        -p 8080:8080 -p 8443:8443 \
        --env ALLOW_EMPTY_PASSWORD=yes \
        --env DRUPAL_DATABASE_USER=bn_drupal \
        --env DRUPAL_DATABASE_PASSWORD=bitnami \
        --env DRUPAL_DATABASE_NAME=bitnami_drupal \
        --env DRUPAL_SKIP_BOOTSTRAP=no \
        --env BITNAMI_DEBUG=true \
        --network drupal-network \
        nycedcmis/bitnami-drupal-custom-theme
```
4. Once the site was up and running I installed and set my custom theme as the default (verified it was working)
5. Exported the db from within the mariadb container: `mysqldump -uroot bitnami_drupal > /tmp/d9.sql`
6. Copied that dump to the host with `docker cp`
7. Destroyed both the running drupal and mariadb containers
8. Started a new mariadb container but this time ensuring it uses the dump taken from the previous run:
```
docker run --rm -d --name mariadb \
	--env ALLOW_EMPTY_PASSWORD=yes \
	--env MARIADB_USER=bn_drupal \
	--env MARIADB_PASSWORD=bitnami \
	--env MARIADB_DATABASE=bitnami_drupal \
	-v $HOME/temp/drupaldb:/docker-entrypoint-initdb.d \
	--network drupal-network \
	bitnami/mariadb:latest
```
9. Rerunning the drupal container with the custom image and skipping the bootstrap process:
```
docker run --rm --name drupal \
        -p 8080:8080 -p 8443:8443 \
        --env ALLOW_EMPTY_PASSWORD=yes \
        --env DRUPAL_DATABASE_USER=bn_drupal \
        --env DRUPAL_DATABASE_PASSWORD=bitnami \
        --env DRUPAL_DATABASE_NAME=bitnami_drupal \
        --env DRUPAL_SKIP_BOOTSTRAP=yes \
        --env BITNAMI_DEBUG=true \
        --network drupal-network \
        nycedcmis/bitnami-drupal-custom-theme
```
10. Verified that the site is using the new custom theme by default

Again, I didn't see any issues except for a 404 because the custom theme doesn't specify a logo.  Now this is a very simple scenario, to properly test this I'll need to have access to your custom db and site or at least some more info on what the site is using (modules, themes, etc.).

alemorcuq commented:
This is what I'm doing. I'm using a `docker-compose.yml` file that deploys MariaDB and Drupal 9 with your fix. No custom themes at all. Only MariaDB has persistence:

```
version: '2'
services:
  mariadb:
    image: 'docker.io/bitnami/mariadb:10.5-debian-10'
    environment:
      - ALLOW_EMPTY_PASSWORD=yes
      - MARIADB_USER=bn_drupal
      - MARIADB_PASSWORD=bitnami
      - MARIADB_DATABASE=bitnami_drupal
    volumes:
      - 'mariadb_data:/bitnami/mariadb'
  drupal:
    image: 'alemorcuq/drupal:9-fix'
    ports:
      - '80:8080'
      - '443:8443'
    environment:
      - DRUPAL_DATABASE_HOST=mariadb
      - DRUPAL_DATABASE_PORT_NUMBER=3306
      - DRUPAL_DATABASE_USER=bn_drupal
      - DRUPAL_DATABASE_NAME=bitnami_drupal
      - DRUPAL_DATABASE_PASSWORD=bitnami
      - DRUPAL_SKIP_BOOTSTRAP=no
      - ALLOW_EMPTY_PASSWORD=yes
    depends_on:
      - mariadb
volumes:
  mariadb_data:
    driver: local
```

After the application is launched, I bring down the containers with `docker-compose down` and then modify the file to enable the `DRUPAL_SKIP_BOOTSTRAP` variable. Then, I bring the containers up again with `docker-compose up`.

In the logs, I can see that MariaDB is using the persisted data, as expected:

```
mariadb_1  | mariadb 14:45:42.67 INFO  ==> Using persisted data
```

I can also see that the Drupal configuration is being skipped:

```
drupal_1   | drupal 14:45:48.21 INFO  ==> An already initialized Drupal database was provided, configuration will be skipped
```

But then when I access the site it returns 404s for the CSS files:

```
drupal_1   | 172.30.0.1 - - [31/Mar/2021:14:45:54 +0000] "GET / HTTP/1.1" 200 2382
drupal_1   | 172.30.0.1 - - [31/Mar/2021:14:45:54 +0000] "GET /sites/default/files/css/css_3V6J_r_PTht_o1AwEAK5H1GPwWEaEBu0a2228ToEXnc.css HTTP/1.1" 404 249
drupal_1   | 172.30.0.1 - - [31/Mar/2021:14:45:54 +0000] "GET /sites/default/files/css/css_v18WUm75dSk_A-DeLZcRjivGVxxUXl9PCwLE_JJpyIk.css HTTP/1.1" 404 249
drupal_1   | 172.30.0.1 - - [31/Mar/2021:14:45:54 +0000] "GET /sites/default/files/css/css_Z5jMg7P_bjcW9iUzujI7oaechMyxQTUqZhHJ_aYSq04.css HTTP/1.1" 404 249
drupal_1   | 172.30.0.1 - - [31/Mar/2021:14:45:55 +0000] "GET /core/themes/bartik/logo.svg HTTP/1.1" 200 3810
drupal_1   | 172.30.0.1 - - [31/Mar/2021:14:45:55 +0000] "GET /core/misc/favicon.ico HTTP/1.1" 200 5430
```

ChexWarrior commented:
@alemorcuq  Was able to reproduce the same error using that setup, I'll investigate what the issue could be.

alemorcuq commented:
Looks like clearing the Drupal cache fixes it, yes. I think we should execute `drush cache-rebuild` automatically so this works out of the box, don't you think?

ChexWarrior commented:
@alemorcuq Agreed, I've updated the code so the cache is always cleared regardless if your skipping bootstrap or not.

marcosbc commented on 8/debian-10/rootfs/opt/bitnami/scripts/libdrupal.sh:
```diff
@@ -280,6 +289,27 @@ drupal_site_install() {
     fi
 }
 
+drupal_create_config_directory() {
```
We should be describing all new functions with a top comment, like we are doing throughout the file. For instance:

```bash
########################
# Drupal remove duplicated database block from settings file
# Globals:
#   *
# Arguments:
#   None
# Returns:
#   None
#########################
```

ChexWarrior replied:
Added to drupal_create_config_directory

marcosbc commented on 8/debian-10/rootfs/opt/bitnami/scripts/libdrupal.sh:
```diff
@@ -143,13 +143,22 @@ drupal_initialize() {
-            drupal_flush_cache
         else
             info "An already initialized Drupal database was provided, configuration will be skipped"
+			if is_empty_value "$DRUPAL_DATABASE_TLS_CA_FILE"; then
```
The indentation of the newly added changes seem weird. Note that don't use tabs in shell scripts. Could you fix it?

marcosbc commented on 8/debian-10/rootfs/opt/bitnami/scripts/libdrupal.sh:
```diff
@@ -143,13 +143,22 @@ drupal_initialize() {
+				drupal_set_database_ssl_settings
+			fi
+			drupal_create_config_directory "$DRUPAL_BASE_DIR/$DRUPAL_CONFIG_SYNC_DIR"
+			drupal_conf_set "\$settings['config_sync_directory']" "$DRUPAL_CONFIG_SYNC_DIR" no
```
The default value for the last parameter of `drupal_conf_set` is `no`, so I would recommend not to specify it, in accordance with what we are doing.

Also, if you could add a comment indicating why it is needed to set the `config_sync_directory` and `hash_salt` values, it would be great for anyone not familiar with the internals of the Drupal application.

marcosbc commented on 8/debian-10/rootfs/opt/bitnami/scripts/libdrupal.sh:
```diff
@@ -280,6 +289,27 @@ drupal_site_install() {
 }
 
+drupal_create_config_directory() {
+	if [[ -z $DRUPAL_CONFIG_SYNC_DIR ]]; then
```
We have some common functions for improving the readability of scripts, such as `is_empty_value`, `generate_random_value` or `ensure_dir_exists`. To improve the readability of this function you could rewrite it as:

```bash
local config_sync_dir="${DRUPAL_CONFIG_SYNC_DIR:-}"
if is_empty_value "$config_sync_dir"; then
    config_sync_dir="${DRUPAL_BASE_DIR}/sites/default/files/config_$(generate_random_string -t alphanumeric -c 16)"
fi
ensure_dir_exists "$config_sync_dir"
```

That way, there is no dependency with Drush and you don't need to set a false value.

marcosbc commented on 8/debian-10/rootfs/opt/bitnami/scripts/libdrupal.sh:
```diff
@@ -280,6 +289,27 @@ drupal_site_install() {
 }
 
+drupal_create_config_directory() {
+	if [[ -z $DRUPAL_CONFIG_SYNC_DIR ]]; then
+		# In order to use Drush commands we need to have a value for
+		# config_sync_directory or it will throw an error so we add a
+		# temp value here first
+		drupal_conf_set "\$settings['config_sync_directory']" "temp" no
+		DRUPAL_CONFIG_SYNC_DIR="sites/default/files/config_$(drush eval "echo(Drupal\Component\Utility\Crypt::randomBytesBase64(55))")"
+		debug_execute mkdir "$DRUPAL_BASE_DIR/$DRUPAL_CONFIG_SYNC_DIR"
+	else
+		debug_execute mkdir "$@"
+	fi
```
We have some common functions for improving the readability of scripts, such as `is_empty_value`, `generate_random_string` or `ensure_dir_exists`. To improve the readability of this function you could rewrite it as:

```suggestion
    local config_sync_dir="${DRUPAL_CONFIG_SYNC_DIR:-}"
    if is_empty_value "$config_sync_dir"; then
        config_sync_dir="${DRUPAL_BASE_DIR}/sites/default/files/config_$(generate_random_string -t alphanumeric -c 16)"
    fi
    ensure_dir_exists "$config_sync_dir"
```

That way, there is no dependency with Drush and you don't need to set a false value.

marcosbc commented on 8/debian-10/rootfs/opt/bitnami/scripts/libdrupal.sh:
```diff
@@ -280,6 +289,27 @@ drupal_site_install() {
+}
+
+drupal_set_hash_salt() {
+	if [[ -z $DRUPAL_HASH_SALT ]]; then
+		drupal_conf_set "\$settings['hash_salt']" "$(drush eval "echo(Drupal\Component\Utility\Crypt::randomBytesBase64(55))")" no
+	else
+		drupal_conf_set "\$settings['hash_salt']" "$DRUPAL_HASH_SALT" no
+	fi
```
In the same way than before, this could be improved as follows:

```suggestion
    local hash_salt="${DRUPAL_HASH_SALT:-}"
    if is_empty_value "$hash_salt"; then
        hash_salt="$(generate_random_string -t alphanumeric -c 32)"
    fi
    drupal_conf_set "\$settings['hash_salt']" "$hash_salt"
```

marcosbc reviewed (changes requested):
Thanks so much for this contribution! Please check my comments. My main requests are oriented towards code legibility and consistency, but the underlying changes look fine.

marcosbc commented on 8/debian-10/rootfs/opt/bitnami/scripts/libdrupal.sh:
```diff
@@ -280,6 +294,40 @@ drupal_site_install() {
+        config_sync_dir="${DRUPAL_BASE_DIR}/sites/default/files/config_$(generate_random_string -t alphanumeric -c 16)"
+    fi
+    ensure_dir_exists "$config_sync_dir"
+}
```
```suggestion
    drupal_conf_set "\$settings['config_sync_directory']" "$config_sync_dir"
}
```

marcosbc reviewed (changes requested):
Just noticed that this may not work in all cases. I've proposed fix in comments. Could you check it? Thanks!

ChexWarrior commented:
@marcosbc I've addressed all the changes you requested, please let me know if there is anything else you'd like me to update.

marcosbc reviewed (approved):
LGTM, thanks a lot for this contribution!

alemorcuq commented:
Thank you very much for your contribution, @ChexWarrior! And sorry again for missing your ping for review.

ChexWarrior commented:
No worries @alemorcuq, thanks to everyone for helping get this finished and merged.
